File: bash-projects/ch01/filereader.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>

int main(int argc, char** argv) {

    std::string filename {};

    if (argc != 2) {
        std::cout << "Enter filename: ";
	std::getline(std::cin, filename);
    } 

    // file stream open
    std::fstream file(filename.c_str(), std::ios::in);
    // test if the file opening fails
    if (file.fail()) {
	std::cout << "Fail doesn't exist.";
	std::exit(-1);
     }
    // read the file if it exist
    while(!file.eof()) {
	    std::cout << file;
    }
    file.close();

    return 0;
}

File: bash-projects/ch01/filereader.py
#!/usr/bin/env python3

import pathlib


def main(filename: str) -> None:
    """check if the file exist
    if it does read it"""
    file = pathlib.Path(filename)
    if file.is_file():
        with open(file) as f:
            for line in f.readlines():
                print(line.strip())
    else:
        print(f"{file} doesn't exist.")
        main(input('Enter a valid file: '))


if __name__ == '__main__':
    """call main function"""
    import sys

    filename = ''
    args = sys.argv
    if len(args) != 2:
        filename = input('Enter a filename: ')
    else:
        filename = args[1]

    main(filename=filename)
